strawberryfields/apps/train/parametrization.py

```diff
@@ -0,0 +1,406 @@
+# Copyright 2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Contains the :class:`~.VGBS` (variational GBS) class which provides a trainable parametrization
+of the GBS probability distribution.
+"""
+from typing import Optional
+
+import numpy as np
+import thewalrus.samples
+from thewalrus._hafnian import reduction
+from thewalrus._torontonian import tor
+from thewalrus.quantum import Qmat, Xmat
+from thewalrus.quantum import find_scaling_adjacency_matrix as rescale
+from thewalrus.quantum import find_scaling_adjacency_matrix_torontonian as rescale_tor
+from thewalrus.quantum import photon_number_mean_vector, pure_state_amplitude
+
+import strawberryfields as sf
+
+
+def rescale_adjacency(A: np.ndarray, n_mean: float, threshold: bool) -> np.ndarray:
+    """Rescale an adjacency matrix so that it can be mapped to GBS.
+
+    An adjacency matrix must have singular values not exceeding one if it can be mapped to GBS.
+    Arbitrary adjacency matrices must first be rescaled to satisfy this condition.
+
+    This function rescales an input adjacency matrix :math:`A` so that the corresponding gaussian
+    state has:
```

Comment:
Would it be worth mentioning that this involves an optimization task ([which might potentially fail](https://github.com/XanaduAI/thewalrus/blob/cf7a19e71f6df5e8017bc21625fa3fb8af6cc5c8/thewalrus/quantum.py#L661))?

Reply:
> @antalszava : I think that error message in The Walrus is unnecesary. One can prove that the root one is looking for a) exists and b) will be found using something like Newton-Raphson. That error message has never been displayed.

Adding this quote here from @nquesada - I think it's ok here not to mention it.